pandas/core/indexing.py

```diff
@@ -146,7 +146,8 @@ def _get_setitem_indexer(self, key):
             return self._convert_tuple(key, is_setter=True)
 
         axis = self.obj._get_axis(0)
-        if isinstance(axis, MultiIndex):
+
+        if isinstance(axis, MultiIndex) and not is_integer(key):
```

Comment:
use ``self.name != 'iloc'``, and use a tests that fails this (IOW a MultiIndex with an integer first level that does not match the positions)